### PR TITLE
feat: add ERC5564Announcer contract and tests

### PR DESCRIPTION
**Pull request by mds1**
See [ERC-5564](https://eips.ethereum.org/EIPS/eip-5564) for the spec. The contract is super simple so we really only have one test, if you have ideas for more tests let me know.

This PR also turns off the metadata hash, that way the resulting bytecode is independent of the machine it's compiled on.

### CODE REVIEW

**Comment by mds1** [comment]
Current dependencies on/for this PR:
* main
  * **PR #2** <a href="https://app.graphite.dev/github/pr/ScopeLift/stealth-address-erc-contracts/2" target="_blank">Graphite</a>  👈
    * **PR #3** <a href="https://app.graphite.dev/github/pr/ScopeLift/stealth-address-erc-contracts/3" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ScopeLift/stealth-address-erc-contracts/2?utm_source=stack-comment).

File: test/Counter.t.sol [hunk 1]
```diff
@@ -1,26 +1,12 @@
...
 pragma solidity 0.8.20;
 
-import "forge-std/Test.sol";
+import {Test} from "forge-std/Test.sol";
 import {Deploy} from "script/Deploy.s.sol";
-import {Counter} from "src/Counter.sol";
+import {ERC5564Announcer} from "src/ERC5564Announcer.sol";
 
-contract CounterTest is Test, Deploy {
+contract ERC5564AnnouncerTest is Test, Deploy {
```
Comment (alexkeating) [inline test/Counter.t.sol]: Should this file be removed as the code is duplicated in `ERC5564Announcer.sol`

**Review by alexkeating — COMMENTED** [review]
Looks good, I think one of the test files can be removed

File: test/Counter.t.sol (hunk 1 above)
Comment (mds1) [reply test/Counter.t.sol]: Ah thanks, removed in d36a4e6

**Review by alexkeating — APPROVED** [review]

**Review by alexkeating — APPROVED** [review]

**Review by jferas — APPROVED** [review]
Looks good

File: src/interfaces/IERC5564Announcer.sol [hunk 2]
```diff
@@ -0,0 +1,39 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.0;
+
+/// @dev Interface for calling the `ERC5564Announcer` contract, which emits an `Announcement` event
+/// to broadcast information about a transaction involving a stealth address. See
+/// [ERC-5564](https://eips.ethereum.org/EIPS/eip-5564) to learn more.
+interface IERC5564Announcer {
+  /// @dev Emitted when something is sent to a stealth address.
+  /// @param schemeId The applied stealth address scheme (such as secp25k1).
```
Comment (apbendi) [inline src/interfaces/IERC5564Announcer.sol]: nit: Should probably make it clear it's an indentifier, i.e. something like "The identifier corresponding to the applied stealth address scheme, such as 1 for secp25k1, as defined by appropriate extension EIPs"

File: src/interfaces/IERC5564Announcer.sol [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  /// @param stealthAddress The computed stealth address for the recipient.
+  /// @param caller The caller of the `announce` function that emitted this event.
+  /// @param ephemeralPubKey Ephemeral public key used by the sender.
```
Comment (apbendi) [inline src/interfaces/IERC5564Announcer.sol]: nit: should this clarify "used by the sender *to encrypt the Announcement metadata*"

File: script/Deploy.s.sol [hunk 4]
```diff
@@ -3,14 +3,14 @@
 
 pragma solidity 0.8.20;
 
-import "forge-std/Script.sol";
-import {Counter} from "src/Counter.sol";
+import {Script} from "forge-std/Script.sol";
+import {ERC5564Announcer} from "src/ERC5564Announcer.sol";
 
 contract Deploy is Script {
```
Comment (apbendi) [inline script/Deploy.s.sol:9]: Is it worth writing a deploy script like this, and integrating it into the tests, when we know the real contract MUST be deployed according to Nick's method at a pre-computed address?

**Review by apbendi — COMMENTED** [review]
Left some minor comments. Off to a good start though!

File: src/interfaces/IERC5564Announcer.sol (hunk 2 above)
Comment (mds1) [reply src/interfaces/IERC5564Announcer.sol]: Updated in e204775. Note that the scheme IDs are directly defined in EIP-5564, not extension EIPs

File: src/interfaces/IERC5564Announcer.sol (hunk 3 above)
Comment (mds1) [reply src/interfaces/IERC5564Announcer.sol]: Updated in e204775. Note that unlike umbra we're not encrypting any data in the announcement, we just use the shared secret generated from the ephemeral pubkey to directly generate the stealth address

File: script/Deploy.s.sol (hunk 4 above)
Comment (mds1) [reply script/Deploy.s.sol:9]: Was chatting with @nerolation about this and current plan is to deploy via the default create2 deployer used by forge, which is [this one](https://github.com/Arachnid/deterministic-deployment-proxy). The rationale is:
- Nick's method requires not signing over chain ID, but some chains (e.g. Celo) require you to sign over chain ID, so it'd be impossible to deploy on that chain.
- Similarly, it requires hardcoding a gas price and gas limit but some chains (e.g. Arbitrum) have different gas limits/prices than other chains, so now you'd need a super high gas limit which isn't great
- Therefore deploying on those chains would require the chains to administratively place the ERC contract code there, which isn't likely.
- But, both of those chains already administratively placed Nick's deterministic deployer at the expected address. In other words, for chains where Nick's method doesn't work, the contract needs to be placed there, and you're more likely to get them to place (and some already have) a generic create2 deployer than a specific EIP's contracts.

Given that plan, I do think it makes to handle the deploy script this way.

**Review by apbendi — APPROVED** [review]
Looks great, thanks!

**Comment by mds1** [comment]
@mds1 started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/ScopeLift/stealth-address-erc-contracts/2).

**Comment by mds1** [comment]
@mds1 merged this pull request with [Graphite](https://app.graphite.dev/github/pr/ScopeLift/stealth-address-erc-contracts/2).

**Pull request merged**